README
TGC is the TCP Gender Changer.

Copyright (C) 2008-2014 Faraz V (faraz@fzv.ca)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

Disclaimer:
This program is provided with no warranty of any kind, either expressed or
implied.  It is the responsibility of the user (you) to fully research and
comprehend the usage of this program.  As with any tool, it can be misused,
either intentionally or unintentionally.
THE AUTHOR(S) IS(ARE) NOT RESPONSIBLE FOR ANYTHING YOU DO WITH THIS PROGRAM
or anything that happens because of your use (or misuse) of this program.


THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
SUCH DAMAGE.



You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


* Technical details:

	tgcd is a simple Unix network utility to extend the accessibility of TCP/IP based network services
	beyond firewalls.
	This can also be used by network analysts and security experts for penetration testing and analyze 
	the security of their network.

	This program has 3 different modes:

	- CC (ConnectConnect)
	- LL (ListenListen)
	- PF (PortForwarder)

	CC and LL nodes are used together to provide access to a service inside a Local 
	Area Network to the external network. PF node is, however, only a simple
	port forwarder.

	CC node resides in the LAN and connect to the LL (on the external network), and
	once a client connects to LL, then CC gets notified and establishes a connection
	to the server on the LAN.

	This program complies with host_access (aka libwrap), to control its accessibility.

	The idea is nothing new and there has been many implementation of it available already, 
	namely:

 	- revinetd (written by Steven M. Gill and Alexandre Carmel-Veilleux)
		available at: http://revinetd.sourceforge.net

	- An implementation using "socat" (an excellent network utility):
		http://www.dest-unreach.org/socat/doc/socat-gender.txt

	- FireWall Tunnel, available at: http://sourceforge.net/projects/firewall-tunnel/

	- An artile explaining the idea: 
		http://www.bandwidthco.com/whitepapers/netforensics/tcpip/TCP-IP%20Gender%20Changer.pdf

	There might be other implementations using netcat or other network utilities.


	For more information and examples on how to use this utility please refer to its
	man page (tgcd(1)).
	
	This software is provided only as a tool for network security analysts and
	penetration testers. We do not take any responsibility for any damage that
	can be done by using or misusing this software. Use it responsibilly and 
	at your own risk (read the disclaimer above)


	NOTE:  This software has only been tested on the following platforms:

	Linux 2.2.x (x86), 2.4.x (x86), 2.6.x (x86, ARM9, AVR32), Cygwin (1.5.x), NetBSD 2.1.0

	Please feel free to send us your comments or any patches you may have to get tgc 
	working on other platforms.





ChangeLog
2014-07-06: Fixed a bug which caused incomplete data transfer in some HTTP connections (1.0.2)
2013-11-06: Added an argument so that an external filter could be called upon accepting connections to permit or reject the connection (1.0.1)
2008-12-18: Fixed a bug where if both CC and LL used "fork" method the child processes could stay on even after the connection was dropped (0.98)

2008-12-17: Fixed a bug where CC node if used with "fork" method could create zombie processes

2008-11-29: Added some fault tolerancy and security fixes, such as we drop 
	    an incomming connection from CC if it's not coming from the 
	    same IP as control connection or if no client is waiting (v0.96)

2008-11-24: Added multiple concurrent connection support (v0.95)

2008-11-17: Added support for host_access (aka libwrap)

2008-11-17: Initial code release (v0.90)



tgc.h
/* tgc.h is part of tgc package.
   Copyright (C) 2014	Faraz.V (faraz@fzv.ca)
  
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.
  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
 
Disclaimer:
   This program is provided with no warranty of any kind, either expressed or
   implied.  It is the responsibility of the user (you) to fully research and
   comprehend the usage of this program.  As with any tool, it can be misused,
   either intentionally or unintentionally.
   THE AUTHOR(S) IS(ARE) NOT RESPONSIBLE FOR ANYTHING YOU DO WITH THIS PROGRAM
   or anything that happens because of your use (or misuse) of this program.
  
   THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
   ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
   IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
   ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
   FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
   DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
   OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
   OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
   SUCH DAMAGE.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#ifndef _TGC_H_
#define _TGC_H_

#define TGC_DEFAULT_LOGFILE	""
#define TGC_DEFAULT_LOGLEVEL	0
#define TGC_BUFFER_SIZE 	4096

// time intervals between each connectin attempt to LL
#define  TGC_TIMEOUT	40

#define	CC_SERVER	tgc->node.cc.server
#define CC_SERVER_PORT	tgc->node.cc.server_port
#define LL_HOST		tgc->node.cc.ll_host
#define CC_LL_PORT	tgc->node.cc.ll_port


#define	TGC_COMM_PING	'P'	
#define	TGC_COMM_CCC	'O'
#define TGC_COMM_CLOSE	'C'

#define TGC_METHOD_FORK		'F'
#define TGC_METHOD_SELECT	'S'

// to queue up the incoming connections to pair them up together later
typedef struct _socketq {
	int	sd;
	struct _socketq	*next;
} socket_queue;

// to hold connection pairs
typedef struct _sock_pair_list {
	int	sdi;			//internal (LL-CC)
	int	sdx;			//external (C/S-CC/LL)
	struct _sock_pair_list	*next;
	struct _sock_pair_list	*prev;
} socket_pair_list;

typedef enum { CCNODE, LLNODE, FNODE }	node_type;

typedef struct {
	char 		server[MAX_PATH];
	char		ll_host[MAX_PATH];
	unsigned short	server_port;
	unsigned short  ll_port;
	unsigned short	interval;
	unsigned char	protocol;
	int		control_sd;	// control connection
	socket_queue	*socketq;	//external (C/S-CC/LL)
} ccnode_type;

typedef struct {
	unsigned short	port;
	unsigned short  ll_port;
	unsigned short	timeout;
	unsigned char	protocol;
	int		control_sd;	// control connection
	socket_queue	*socketq;	//external (C/S-CC/LL)
} llnode_type;


typedef struct {
	char 		dst_host[MAX_PATH];
	unsigned short	dst_port;
	unsigned short  port;
	unsigned char	src_protocol;
	unsigned char	dst_protocol;
} fnode_type;


typedef struct {
	node_type	type;
	union {
		ccnode_type	cc;
		llnode_type	ll;
		fnode_type	pf;
	} node;
	int			sdi_accept;
	int			sdx_accept;
	socket_pair_list	*pairs;
	unsigned char   	buf[TGC_BUFFER_SIZE];
	unsigned char		key;
	char   			method;
	char   			filter[MAX_PATH + 1];
} TGC;


typedef struct {
	char	*name;
	char	*addr;
	struct  sockaddr_in *sin;
	int	sd;
} host_info;

/* OP Codes */

/*
Messege Structure:

0: OP Code (1B)
1: Msg. Len (2B)
2: DATA
*/

/*-----------------------------------------*/

/* Error codes  */
#define E_TGC_OK		0
#define E_TGC_NOCANDO		-1
#define E_TGC_BADOPCODE		-2
#define E_TGC_IE		-3
#define E_TGC_READ		-4
#define E_TGC_WRITE		-5
#define E_TGC_END		-6


/* function prototypes */
int tgc_pre_init(TGC *tgc);
int tgc_post_init(TGC *tgc);
void tgc_shutdown(TGC *tgc);

int tgc_run(TGC *tgc);
int tgc_cc(TGC *tgc);
int tgc_ll(TGC *tgc);
int tgc_pf(TGC *tgc);

#endif



utils.h
/* utils.h is part of tgc package
   Copyright (C) 2008	Faraz.V (faraz@fzv.ca)
  
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.
  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
 
Disclaimer:
   This program is provided with no warranty of any kind, either expressed or
   implied.  It is the responsibility of the user (you) to fully research and
   comprehend the usage of this program.  As with any tool, it can be misused,
   either intentionally or unintentionally.
   THE AUTHOR(S) IS(ARE) NOT RESPONSIBLE FOR ANYTHING YOU DO WITH THIS PROGRAM
   or anything that happens because of your use (or misuse) of this program.
 
   THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
   ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
   IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
   ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
   FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
   DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
   OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
   OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
   SUCH DAMAGE.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#ifndef __UTILS_H__
#define __UTILS_H__

#define DEBUG

#ifdef __cplusplus
  extern "C" {
#endif 

#ifndef MAX_PATH
#define  MAX_PATH	254
#endif

void init_log(int dlevel, char *dbg_filename, int daemon);

void close_log(void);

#ifdef __STDC__
void print_log(int dlevel, const char *format_str, ...);
#else
void print_log(va_alist);
#endif

#ifdef __STDC__
void print_log_msg(int dlevel, const char *func, const char *format_str, ...);
#else
void print_log_msg(va_alist);
#endif

#if __GNUC__ > 3
	#ifdef DEBUG
	#define	PRINT_LOG(l, ...)	print_log_msg((l), __FUNCTION__, __VA_ARGS__)
	#else
	#define	PRINT_LOG(l, ...)	
	#endif
#else
	#ifdef DEBUG
	#define	PRINT_LOG(l, args...)	print_log_msg((l), __FUNCTION__, args)
	#else
	#define	PRINT_LOG(l, args...)	
	#endif
#endif

int 	big_endian(void);
int 	strequal(char *s1, char *s2);
void 	strlower(char *s);
void 	strupper(char *s);
void 	string_replace(char *s, char old_str, char new_str);
int	strnum(char *s);
void 	become_daemon(void);
void 	change_user(void);
int 	xioctl(int fd, int request, void *argp);
int 	read_data(int fd, unsigned char *buf, int len);
int 	parse_host(char *hostport, char *host, unsigned short *port);
int 	connect_server(char *host, int port);
int	open_server_socket(int port);
void 	close_connection(int *sd);
RETSIGTYPE sig_cld();
#ifdef HAVE_LIBWRAP
int	peer_ok(char *prog, int sd);
#endif
int 	accept_connection(int sd_accept, struct sockaddr_in *addr, socklen_t *in_addrlen);

#ifdef __cplusplus
  }
#endif 

#endif



config.h
/* config.h.  Generated from config.h.in by configure.  */
/* config.h.in.  Generated from configure.in by autoheader.  */

/* Define to 1 if you have the <arpa/inet.h> header file. */
#define HAVE_ARPA_INET_H 1

/* Define to 1 if you have the <asm/types.h> header file. */
#define HAVE_ASM_TYPES_H 1

/* Define to 1 if you have the <bits/sockaddr.h> header file. */
#define HAVE_BITS_SOCKADDR_H 1

/* Define to 1 if you have the <fcntl.h> header file. */
#define HAVE_FCNTL_H 1

/* Define to 1 if you have the `fork' function. */
#define HAVE_FORK 1

/* Define to 1 if you have the `gethostbyname' function. */
#define HAVE_GETHOSTBYNAME 1

/* Define to 1 if you have the `gettimeofday' function. */
#define HAVE_GETTIMEOFDAY 1

/* Define to 1 if you have the `inet_ntoa' function. */
#define HAVE_INET_NTOA 1

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define to 1 if you have the `nsl' library (-lnsl). */
#define HAVE_LIBNSL 1

/* Define to 1 if you have the `wrap' library (-lwrap). */
#define HAVE_LIBWRAP 1

/* Define to 1 if your system has a GNU libc compatible `malloc' function, and
   to 0 otherwise. */
#define HAVE_MALLOC 1

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H 1

/* Define to 1 if you have the `memset' function. */
#define HAVE_MEMSET 1

/* Define to 1 if you have the <netdb.h> header file. */
#define HAVE_NETDB_H 1

/* Define to 1 if you have the <netinet/in.h> header file. */
#define HAVE_NETINET_IN_H 1

/* Define to 1 if you have the `select' function. */
#define HAVE_SELECT 1

/* Define to 1 if you have the `socket' function. */
#define HAVE_SOCKET 1

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the `strcasecmp' function. */
#define HAVE_STRCASECMP 1

/* Define to 1 if you have the `strdup' function. */
#define HAVE_STRDUP 1

/* Define to 1 if you have the `strerror' function. */
#define HAVE_STRERROR 1

/* Define to 1 if you have the `strftime' function. */
#define HAVE_STRFTIME 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the <syslog.h> header file. */
#define HAVE_SYSLOG_H 1

/* Define to 1 if you have the <sys/file.h> header file. */
#define HAVE_SYS_FILE_H 1

/* Define to 1 if you have the <sys/ioctl.h> header file. */
#define HAVE_SYS_IOCTL_H 1

/* Define to 1 if you have the <sys/param.h> header file. */
#define HAVE_SYS_PARAM_H 1

/* Define to 1 if you have the <sys/select.h> header file. */
#define HAVE_SYS_SELECT_H 1

/* Define to 1 if you have the <sys/socket.h> header file. */
#define HAVE_SYS_SOCKET_H 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/time.h> header file. */
#define HAVE_SYS_TIME_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have <sys/wait.h> that is POSIX.1 compatible. */
#define HAVE_SYS_WAIT_H 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Define to 1 if you have the `vfork' function. */
#define HAVE_VFORK 1

/* Define to 1 if you have the <vfork.h> header file. */
/* #undef HAVE_VFORK_H */

/* Define to 1 if you have the `wait3' system call. Deprecated, you should no
   longer depend upon `wait3'. */
#define HAVE_WAIT3 1

/* Define to 1 if `fork' works. */
#define HAVE_WORKING_FORK 1

/* Define to 1 if `vfork' works. */
#define HAVE_WORKING_VFORK 1

/* Name of package */
#define PACKAGE "tgcd"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT "faraz@fzv.ca"

/* Define to the full name of this package. */
#define PACKAGE_NAME "tgcd"

/* Define to the full name and version of this package. */
#define PACKAGE_STRING "tgcd 1.0.2"

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME "tgcd"

/* Define to the version of this package. */
#define PACKAGE_VERSION "1.0.2"

/* Define as the return type of signal handlers (`int' or `void'). */
#define RETSIGTYPE void

/* Define to the type of arg 1 for `select'. */
#define SELECT_TYPE_ARG1 int

/* Define to the type of args 2, 3 and 4 for `select'. */
#define SELECT_TYPE_ARG234 (fd_set *)

/* Define to the type of arg 5 for `select'. */
#define SELECT_TYPE_ARG5 (struct timeval *)

/* Define to 1 if the `setpgrp' function takes no argument. */
#define SETPGRP_VOID 1

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS 1

/* Define to 1 if you can safely include both <sys/time.h> and <time.h>. */
#define TIME_WITH_SYS_TIME 1

/* Version number of package */
#define VERSION "1.0.2"

/* Define to empty if `const' does not conform to ANSI C. */
/* #undef const */

/* Define to rpl_malloc if the replacement function should be used. */
/* #undef malloc */

/* Define to `int' if <sys/types.h> does not define. */
/* #undef pid_t */

/* Define to `unsigned int' if <sys/types.h> does not define. */
/* #undef size_t */

/* Define as `fork' if `vfork' does not work. */
/* #undef vfork */


tgcd.c
/* tgcd.c is part of tgc package.
   Copyright (C) 2014	Faraz.V (faraz@fzv.ca)
  
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.
  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

Disclaimer:
   This program is provided with no warranty of any kind, either expressed or
   implied.  It is the responsibility of the user (you) to fully research and
   comprehend the usage of this program.  As with any tool, it can be misused,
   either intentionally or unintentionally.
   THE AUTHOR(S) IS(ARE) NOT RESPONSIBLE FOR ANYTHING YOU DO WITH THIS PROGRAM
   or anything that happens because of your use (or misuse) of this program.

   THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
   ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
   IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
   ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
   FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
   DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
   OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
   OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
   SUCH DAMAGE.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include "includes.h"
#include "utils.h"
#include "tgc.h"

int initial_uid = 0;
int initial_gid = 0;

int	timeout = 250;

const char *program_name;

#ifdef HAVE_FORK
// We try to be a daemon by default 
int	is_daemon = 1;
#else
int	is_daemon = 0;
#endif

TGC	tgc;

/*------------------------------------------------------------------------------
   Prints version info.
--------------------------------------------------------------------------------*/
void print_version(void)
{
	printf("TCP Gender Changer, V%s Copyright (C) 2014 Faraz.V (faraz@fzv.ca)\n", VERSION);
}

/*------------------------------------------------------------------------------
   Prints usage info.
--------------------------------------------------------------------------------*/
void print_usage(int exit_code)
{
	print_version();
	printf("Usage: %s { -C | -L | -F }  options ... \n\n", PACKAGE);
	printf("Options are:\n\n");
	

	printf(" ConnectConnect mode: %s -C -s host:port -c host:port [-i n ] [-k n ] [ common options ]\n", PACKAGE);
	printf(" ConnectConnect :\n");
	printf(" -C, --ccnode 		    Become a CC (ConnectConnect) node.\n");
	printf(" -s, --server host:port     The host and port of the actual server\n");
	printf(" -c, --llhost host:port     The host and port of the ListenListen node.\n");
	printf(" -i, --interval seconds     Time interval to periodically report to LL (default: %ds).\n", TGC_TIMEOUT);
	printf(" -k, --key number	    Poorman's encryption (0-255, default: 0, means no encryption)\n");
	
	printf("\n");
	printf(" ListenListen mode: %s -L -p port  -q port [-t n] [-k n ] [ common options ...]\n", PACKAGE);
	printf(" -L, --llnode		    Become a LL (ListenListen) node.\n");
	printf(" -q, --llport number 	    The port to listen on for incomming connection from a CC node\n");
	printf(" -p, --port number 	    The port to listen on for incomming actual client connection\n");
	printf(" -t, --timeout seconds      Inactivity timeout during connection with CC node (default: %ds).\n",
	       (int)(TGC_TIMEOUT * 1.5));
	printf(" -k, --key number 	    Poorman's encryption (0-255, default: 0, means no encryption)\n");
	
	printf("\n");
	printf(" PortForwarder mode: %s -F -p port -s host:port [ common options ... ]\n", PACKAGE);
	printf(" -F, --lcnode		    Become a ListenConnect node, i.e. just a simple port forwarder\n");
	printf(" -p, --port  number 	    The port to listen on for incomming actual client connection\n");
	printf(" -s, --server host:port     The host and port of the actual server or a LL node\n");

	printf("\n");
	printf("Common options:\n");
        printf(" -m, --method { f | s }	    f: Fork  s: Select (default: s)\n");
        printf(" -f, --filter filter	    Optional argument to run filter on new connections, IP passed as argument\n");
        printf(" -l, --log file 	    Write logs to file. (default:'%s')\n", TGC_DEFAULT_LOGFILE);
	printf(" -g, --level number 	    Log level detail (default:%d).\n", TGC_DEFAULT_LOGLEVEL);
	printf(" -n, --nodaemon             Do not become daemon\n");
	printf(" -h, --help		    Display this.\n");
	printf(" -v, --version		    Display version number.\n\n");

	exit(exit_code);
}

/*-----------------------------------------------------------------------------
   Shutdown the server
------------------------------------------------------------------------------*/
RETSIGTYPE shutdown_server(int sig)
{
	PRINT_LOG(1, "Shutting down (signal:%d)", sig);
	tgc_shutdown(&tgc);
	exit(0);
}

/*-----------------------------------------------------------------------------
   Initialize the server data structures and variables.
------------------------------------------------------------------------------*/
void init_server(void)
{
	big_endian();

	if (tgc_post_init(&tgc)<0) {
		PRINT_LOG(1, "Failed to initialize!");
		print_usage(2);
	}

	initial_uid = geteuid();
	initial_gid = getegid();

	signal(SIGPIPE, SIG_IGN);

	signal(SIGTERM, shutdown_server);
	signal(SIGINT, shutdown_server);
	signal(SIGQUIT, shutdown_server);
	signal(SIGHUP, shutdown_server);

	if (tgc.method==TGC_METHOD_FORK) 
		signal(SIGCLD, sig_cld);
}

/*-----------------------------------------------------------------------------
  The Main function !
------------------------------------------------------------------------------*/
int main(int argc,char *argv[])
{
	int	loglevel     = TGC_DEFAULT_LOGLEVEL;
	int	next_option;
	extern	char *optarg;	/* getopt */
	extern  int  optind;	/* getopt */
	char	logfilename[MAX_PATH+1] = {0};
	char	temp[MAX_PATH] = {0};
	int	ntemp=0, rc=0;
	struct stat filter_stat;
	
	/* short options */
	const char *short_options = "Cs:c:i:Lq:p:t:Fk:m:f:l:g:nhv";
	
	/* long options */
	const struct option long_options[] = {
		{"ccnode",	0, NULL, 'C'},
		{"server",	1, NULL, 's'},
		{"llhost",	1, NULL, 'c'},
		{"interval",	1, NULL, 'i'},

		{"llnode",	0, NULL, 'L'},
		{"llport",	1, NULL, 'q'},
		{"port",	1, NULL, 'p'},
		{"timeout",	1, NULL, 't'},

		{"lcnode",	0, NULL, 'F'},

		{"key",		1, NULL, 'k'},

		{"method",	1, NULL, 'm'},
		{"filter",	1, NULL, 'f'},

		{"log",		1, NULL, 'l'},
		{"level",	1, NULL, 'g'},
		{"nodaemon",	0, NULL, 'n'},
		{"help",	0, NULL, 'h'},
		{"version",	0, NULL, 'v'},
		{NULL, 		0, NULL, 0}
	};

	strncpy(logfilename, TGC_DEFAULT_LOGFILE, MAX_PATH);
	tgc_pre_init(&tgc);

	program_name = argv[0];

	do {
		next_option = getopt_long(argc, argv, short_options, long_options, NULL);

		switch (next_option) {
			case 'C':
				tgc.type=CCNODE;
				break;
			case 's':
				strncpy(temp, optarg, MAX_PATH);
				if (tgc.type==FNODE)
                                	parse_host(temp, tgc.node.pf.dst_host, &(tgc.node.pf.dst_port) );
				else
                                	parse_host(temp, tgc.node.cc.server, &(tgc.node.cc.server_port) );
				break;
			case 'c':
				strncpy(temp, optarg, MAX_PATH);
                                parse_host(temp, tgc.node.cc.ll_host, &tgc.node.cc.ll_port);
				break;
			case 'L':
				tgc.type=LLNODE;
				break;
			case 'q':
				ntemp = atoi(optarg);
				if (!strnum(optarg) || ntemp<=0 || ntemp>65535) {
					fprintf(stderr, "Invalid port number '%s'\n", optarg);
					exit(2);
				}
				tgc.node.ll.ll_port = ntemp;
				break;
			case 'F':
				tgc.type=FNODE;
				break;
			case 'p':
				ntemp = atoi(optarg);
				if (!strnum(optarg) || ntemp<=0 || ntemp>65535) {
					fprintf(stderr, "Invalid port number '%s'\n", optarg);
					exit(2);
				}
				if (tgc.type == FNODE)
					tgc.node.pf.port = ntemp;
				else
					tgc.node.ll.port = ntemp;
				break;
			case 'l':
				strncpy(logfilename, optarg, MAX_PATH);		
				break;
			case 'f':
				strncpy( tgc.filter, optarg, MAX_PATH);
				rc = stat(tgc.filter, &filter_stat);
				if (rc || !S_ISREG(filter_stat.st_mode)) {
					fprintf(stderr, "Invalid filter '%s'\n", optarg);
					exit(3);
				}
				break;
			case 'm':
				strncpy( &(tgc.method), optarg, 1);
				break;
			case 'g':
				ntemp = atoi(optarg);
				if (!strnum(optarg) || ntemp<0) {
					fprintf(stderr, "Invalid loglevel '%s'\n", optarg);
					exit(2);
				}
				loglevel = ntemp;
				break;
			case 'i':
				ntemp = atoi(optarg);
				if (!strnum(optarg) || ntemp<=0 || ntemp>65535) {
					fprintf(stderr, "Invalid interval number '%s'\n", optarg);
					exit(2);
				}
				tgc.node.cc.interval = ntemp;
				break;
			case 't':
				ntemp = atoi(optarg);
				if (!strnum(optarg) || ntemp<=0 || ntemp>65535) {
					fprintf(stderr, "Invalid interval number '%s'\n", optarg);
					exit(2);
				}
				tgc.node.ll.timeout = ntemp;
				break;
			case 'k':
				ntemp = atoi(optarg);
				if (!strnum(optarg) || ntemp<0 || ntemp>255) {
					fprintf(stderr, "Invalid key '%s'\n", optarg);
					exit(2);
				}
				tgc.key = ntemp;
				break;
			case 'n':
				is_daemon = 0;
				break;
			case 'h':
				print_usage(0);
				break;
			case 'v':
				print_version();
				exit(0);
				break;
			case -1:
				/* done with options! */
				break;
			case '?':
			default:
				print_usage(1);
		}
	} while (next_option != -1);
	
	/* if there are non-option parameters printn usage and exit! */
	if (optind<argc || argc<=1) 
		print_usage(1);

	init_log(loglevel, logfilename, is_daemon);

	init_server();

#ifdef HAVE_FORK
	if (is_daemon)
		become_daemon();
#endif
	tgc_run(&tgc);

	close_log();
	return 0;
}

